src/Framework/TestCase.php

```diff
@@ -97,8 +97,13 @@ public function runTest($method, array $args = NULL)
 				if (!is_array($res) && !$res instanceof \Traversable) {
 					throw new TestCaseException("Data provider $provider() doesn't return array or Traversable.");
 				}
-				foreach ($res as $set) {
-					$data[] = is_string(key($set)) ? array_merge($defaultParams, $set) : $set;
+				foreach ($res as $setName => $set) {
+					$set = is_string(key($set)) ? array_merge($defaultParams, $set) : $set;
+					if (is_string($setName)) {
+						$data[$setName] = $set;
+					} else {
+						$data[] = $set;
+					}
```

Comment:
@milo Unfortunately I can't because it breaks possibility of multiple data providers. The string key must be unique because it is explicit value however two different data providers without naming returns array with indexes `0, 1, 2, ...` so it results in wrong data set. You can try it on your own ;-)